test/remote/redis_tests.js

```diff
@@ -0,0 +1,240 @@
+  let connections
+
+  before(() => {
+    return P.all([ redis.acquire(), redis.acquire() ])
+      .then(results => connections = results)
+  })
+
+  after(() => {
+    return P.all(connections.map(connection => redis.release(connection)))
+      .then(() => redis.drain())
+      .then(() => redis.clear())
+  })
+
+  it('first connection reads null', () => {
+    return connections[0].get('foo')
+      .then(result => assert.ok(result === null))
+  })
+
+  it('second connection reads null', () => {
+    return connections[1].get('foo')
+      .then(result => assert.ok(result === null))
+  })
+
+  describe('connection.set:', () => {
+    before(() => {
+      return connections[0].set('foo', 'bar')
+    })
+
+    it('first connection reads data', () => {
+      return connections[0].get('foo')
+        .then(result => assert.equal(result, 'bar'))
+    })
+
+    it('second connection reads data', () => {
+      return connections[1].get('foo')
+        .then(result => assert.equal(result, 'bar'))
+    })
+  })
+
+  describe('concurrent sets:', () => {
+    before(() => {
+      return P.all(connections.map((connection, index) => connection.set('foo', `${index}`)))
+    })
+
+    it('data was set', () => {
+      return connections[0].get('foo')
+        .then(result => assert.ok(result === '0' || result === '1'))
+    })
+  })
+
+  describe('connection.update:', () => {
+    before(() => {
+      return connections[0].update('foo', oldValue => `${oldValue}2`)
+    })
+
+    it('first connection reads data', () => {
+      return connections[0].get('foo')
+        .then(result => assert.equal(result[1], '2'))
+    })
+
+    it('second connection reads data', () => {
+      return connections[1].get('foo')
+        .then(result => assert.equal(result[1], '2'))
+    })
+  })
+
+  describe('update non-existent key:', () => {
+    before(() => {
+      return connections[0].update('wibble', () => 'blee')
+    })
+
+    it('data was updated', () => {
+      return connections[1].get('wibble')
+        .then(result => assert.equal(result, 'blee'))
+    })
+  })
+
+  describe('update existing key to falsey value:', () => {
+    before(() => {
+      return connections[0].update('wibble', () => '')
+    })
+
+    it('data was updated', () => {
+      return connections[1].get('wibble')
+        .then(result => assert.ok(result === null))
+    })
+  })
+
+  describe('concurrent updates of the same key:', () => {
+    const errors = []
+    let successIndex
+
+    before(() => {
+      return P.all(
+        connections.map(
+          (connection, index) => {
+            return connection.update('foo', () => `${index}`)
+              .then(() => successIndex = index)
+              .catch(error => errors.push(error))
+          }
+        )
+      )
+    })
+
+    it('one update failed', () => {
+      assert.equal(errors.length, 1)
+      assert.equal(errors[0].message, 'Unspecified error')
+      assert.equal(errors[0].errno, 999)
+    })
```

Comment:
Is there any risk of these two updates running completely sequentially, so that they both succeed and cause sporadic test failures here?  If so, perhaps we could prevent it by having the `getValue` callback do some sort of async spin-wait, waiting for both connections to reach that point before proceeding.

Reply:
I don't think so because all the Redis commands inside `update` are async and it's only the initial `WATCH` command that that needs to be run in order for the test to fail as we expect. So the two watches are always sent immediately and everything else happens asynchronously after that, in whatever order. To wit:

```
~/c/fxa-auth-server (phil/2209) $ git diff
diff --git a/lib/redis-connection.js b/lib/redis-connection.js
index 9b253a6d..bd04fea8 100644
--- a/lib/redis-connection.js
+++ b/lib/redis-connection.js
@@ -37,10 +37,18 @@ module.exports = (log, client) => {

       isUpdating = true

+      console.log('WATCH')
       return client.watchAsync(key)
-        .then(() => client.getAsync(key))
-        .then(getValue)
+        .then(() => {
+          console.log('GET')
+          return client.getAsync(key)
+        })
+        .then(value => {
+          console.log('getValue')
+          return getValue(value)
+        })
         .then(value => {
+          console.log('MULTI')
           const multi = client.multi()

           if (value) {
@@ -52,12 +60,14 @@ module.exports = (log, client) => {
           return multi.execAsync()
         })
         .catch(err => {
+          console.log('UNWATCH')
           client.unwatch()
           log.error({ op: 'redis.update.error', key, err: err.message })
           isUpdating = false
           throw err
         })
         .then(result => {
+          console.log('done')
           isUpdating = false
           if (! result) {
             log.error({ op: 'redis.watch.conflict', key })
~/c/fxa-auth-server (phil/2209) $ mocha test/remote/redis_tests.js --grep="concurrent updates of the same key"

  pool.acquire:
    concurrent updates of the same key:
WATCH
WATCH
GET
GET
getValue
MULTI
getValue
MULTI
done
done
      ✓ one update failed
      ✓ the other update completed successfully


  2 passing (25ms)
```

However, I guess it is possible, in theory, for one of the two watches not to reach the Redis server before all of the commands on the other connection complete. But given that both connections are to the same server, which for local dev and CI is running on localhost anyway, and there's a fair number amount of work to be done after the watch happens, I don't see it happening in reality. Even when the tests are run against prod boxes, I presume we're talking to a Redis instance in the same AWS region so it's not like the traffic has to go a long way.

Having said all that, I'm still happy to jump through some extra hoops if any of what I just wrote does not persuade you.

Reply:
Colour me persuaded :-)

In a threaded world I'd be a lot more suspicious here, but I think the way the js event loop works actually makes this a lot more predictable.  If we do happen to see sporadic failures here, we can always revisit, and it'll probably be obvious what's causing them.